rfq/order.go

```diff
@@ -530,6 +662,11 @@ type OrderHandler struct {
 	// associated asset transaction policies.
 	policies lnutils.SyncMap[SerialisedScid, Policy]
 
+	// htlcToPolicy maps an HTLC circuit key to the policy that applies to
+	// it. We need this map because for failed HTLCs we don't have the RFQ
+	// data available, so we need to cache this info.
+	htlcToPolicy lnutils.SyncMap[models.CircuitKey, Policy]
```

Comment:
If we can manage to do that then all the HTLC states can be nested under the relevant policies.

Reply:
Yes `OrderHandler.fetchPolicy` does find the scid but that's not done based on the `CircuitKey` alone, it also uses the `InterceptedHtlc.InWireCustomRecords` field to extract the scid directly. Unfortunately this field is not available when you get notified for the HTLC cancel event*, so that's why I'm caching this info on this layer.

\*It's not a value that's simply not exposed on the API, but more like data that is not available in the lnd channel link by the time we process the HTLC cancel. To change LND and have that data available by the API directly would involve doing the same kind of HTLC-caching on the LND link level (or something similar), which could get much more complicated.

I def agree that eliminating HTLC related state here would be a win, but can't see a workaround.

Reply:
I'm trying to remember how this work. Are there HTLCs where we can only get the SCID from their custom records? Is the custom records check strictly necessary for some HTLCs?

Reply:
For incoming asset payment HTLCs we inspect custom records. From what I understand and according to @guggero .

So I will approve this PR as is. I will leave merging down to you @GeorgeTsagk . I'll put up a PR to clarify `OrderHandler.fetchPolicy` with comments.

This PR should work as it is either way IMO.

Reply:
Yes, we can only retrieve the SCID from the custom records. The ChanID of the circuit keys correspond to the actual scids of the channels that carry the HTLCs